### PR TITLE
Bug: fix vulnerable regexes to avoid potential ReDoS

### PR DESCRIPTION
**Pull request by yetingli**
The ambiguity of vulnerable regex is eliminated, so that when the fixed regex matches a string, there is only a unique path to match, thereby ensuring that the fixed regex is safer and faster to match.
This related issue addresses #3811

### CODE REVIEW

**Comment by hippo91** [comment]
@yetingli thanks for this PR.
I tried to understand and test the change you propose. I wrote the following code:
```
import re
from time import perf_counter

SPECIAL_WEAK = re.compile(r"^__[^\W_]+\w*__$")
SPECIAL_STRONG = re.compile(r"^__([^\W_]\w*)+__$")

ATTACK = "__"+ "1"*5000 + "!"

BEGIN = perf_counter()
SPECIAL_WEAK.search(ATTACK)
DURATION = perf_counter() - BEGIN
print(f"Weak took {DURATION} seconds!") 

BEGIN = perf_counter()
SPECIAL_STRONG.search(ATTACK)
DURATION = perf_counter() - BEGIN
print(f"Strong took {DURATION} seconds!") 
```
where `SPECIAL_WEAK` is what is inside pylint today and `SPECIAL_STRONG` what you propose.
When executing it, the weak form took 0.22 seconds but strong form is running for a much longer duration.

Can you explain me what i did wrong?
Thanks again.

**Comment by Pierre-Sassoulas** [comment]
I don't think #3811 is really applicable as an attack vector, because you would have to control the code base and change the names of your functions to something really long (5000+ characters?) and then force the victim to launch pylint on it. I don't think one would want to DOS its own CI. Also, there's an easier way to DOS an online CI by launching pylint 100 times instead of one. Seems to me like the basis for an "uneffficient regex goes brrrr" meme :p 

Joking aside, I think that if we change this regex, because of the possibility of an attack vector for DOS, it should be faster or equivalent in the standard use case too.

**Comment by yetingli** [comment]
Hi @hippo91,
I am sorry, there were several typos before.

The repaired regexes are as follows:
`SPECIAL = re.compile(r"^__([^\W_]_*)+__$")`
`PRIVATE = re.compile(r"^__(_*[^\W_])+_?$")`
You can test the following code:
```
import re
from time import perf_counter

SPECIAL_WEAK = re.compile(r"^__[^\W_]+\w*__$")
SPECIAL_STRONG = re.compile(r"^__([^\W_]_*)+__$")

ATTACK = "__"+ "1"*5000 + "!"

BEGIN = perf_counter()
SPECIAL_WEAK.search(ATTACK)
DURATION = perf_counter() - BEGIN
print(f"Weak took {DURATION} seconds!")

BEGIN = perf_counter()
SPECIAL_STRONG.search(ATTACK)
DURATION = perf_counter() - BEGIN
print(f"Strong took {DURATION} seconds!")
```

The weak form took 0.22 seconds and strong took 0.0014  seconds.

**Comment by yetingli** [comment]
> I don't think #3811 is really applicable as an attack vector, because you would have to control the code base and change the names of your functions to something really long (5000+ characters?) and then force the victim to launch pylint on it. I don't think one would want to DOS its own CI. Also, there's an easier way to DOS an online CI by launching pylint 100 times instead of one. Seems to me like the basis for an "uneffficient regex goes brrrr" meme :p
> 
> Joking aside, I think that if we change this regex, because of the possibility of an attack vector for DOS, it should be faster or equivalent in the standard use case too.

Sorry, I am not very familiar with your project. I just use your project to verify the effectiveness of my ReDos detection tool Orz. 
In addition,  5000+ characters is just a special case.
I think even if it is not so easy to trigger a Dos attack, it will affect the efficiency of the operation.
Whether this issue is serious or not is up to you to decide :).

**Comment by Pierre-Sassoulas** [comment]
> I just use your project to verify the effectiveness of my ReDos detection tool Orz.

Makes sense. Well, the name we're using the regex on are typically name chosen by developer for their variables, functions and classes, so probably between 1 to 30 characters with a mean around 15 characters. If we can find a regex that is faster for those cases too, then sure, let's change it :)

**Comment by yetingli** [comment]
> The results of the script seems to be a little faster for the strong form

Thank you for the test!
Yes, because the weak form is ambiguous and the strong one does not exist.
Intuitively,  suppose the given string is `"11"`, there are two possible paths in the weak form to generate this string, while
there is only one path in the strong form.
When the given string is `"111"`, there are three possible paths in the weak form, whereas there is still one
possible path in the strong form. 
This illustrates how the two search spaces differ as the length of the input increases.
You can use a regex visualization tool, for example [Regexplained](http://www.regexplained.co.uk/),  to more intuitively observe the difference between the two.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for the explanation. This seems like something we want to merge, what do you think @hippo91 ? @yetingli Could you squash all your commits into one and amend the message with the explanation you just gave ? You  can also add yourself to the contributors :)

File: pylint/pyreverse/utils.py
Comment (hippo91) [reply pylint/pyreverse/utils.py:54]: @Pierre-Sassoulas lets merge this! More secure and faster: its a great change.
Thanks @yetingli
Please do not forget to add an entry in the ChangeLog and doc/whatsnew/2.6.rst.

**Comment by Pierre-Sassoulas** [comment]
I tried to rebase this and squash before merging in master, but I don't think this branch is modifiable by maintainers. Could you make it editable @yetingli ? Also I think the change should be added in 2.7.rst not 2.6.rst, now that 2.6 is out, I opened #3823 to add the file, what do you think @hippo91 ?

**Comment by hippo91** [comment]
Yes of course! Thanks @Pierre-Sassoulas . Sorry for the mistake @yetingli! It is indeed in 2.7.rst.

**Comment by Pierre-Sassoulas** [comment]
OK, I'm going to merge this tomorrow after #3823 then :)

**Comment by hippo91** [comment]
Ouuupps. Not so sure after all. In fact next version will be 2.6.1 so it make sense to modify 2.6.rst...
I ll try to verify this ASAP. Sorry for this hesitation...

**Comment by Pierre-Sassoulas** [comment]
@yetingli you can git reset --hard on https://github.com/Pierre-Sassoulas/pylint/tree/yetingli-regex :)

**Review by hippo91 — APPROVED** [review]
Thanks @yetingli !

File: pylint/pyreverse/utils.py
Comment (hippo91) [inline pylint/pyreverse/utils.py:53]: I found this comment not very useful. IMH It is more a commit message.

**Review by hippo91 — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
I recreated the MR in #3828 and also added @yetingli in the CONTRIBUTORS.txt, I close but I'll merge the other one as soon as the tests pass. As hippo91 said, thank you for the fix, much appreciated.

**Pull request closed without merging**
